src/main/java/de/tum/in/www1/artemis/service/LearningGoalProgressService.java

```
@@ -0,0 +1,189 @@
+            }
+            else {
+                return;
+            }
+
+            if (learningGoals == null) {
+                // Learning goals couldn't be loaded, the exercise/lecture unit might have already been deleted
+                return;
+            }
+
+            // The participant is normally a single user, but we need to support teams as well
+            participant.getParticipants().forEach(user -> {
+                learningGoals.forEach(learningGoal -> {
+                    updateProgress(learningGoal.getId(), user);
+                });
+            });
+        }
+        catch (Exception e) {
+            logger.error("Exception while updating progress for learning goal", e);
+        }
+    }
+
+    /**
+     * Updates the progress value (and confidence score) of the given learning goal and user
+     * @param learningGoalId The id of the learning goal to update the progress for
+     * @param user The user for which the progress should be updated
+     */
+    private void updateProgress(Long learningGoalId, User user) {
+        var learningGoal = learningGoalRepository.findByIdWithExercisesAndLectureUnitsAndCompletions(learningGoalId).orElse(null);
+
+        if (user == null || learningGoal == null) {
+            // If the user or learning goal no longer exist, there is nothing to do
+            return;
+        }
+
+        var learningGoalProgress = learningGoalProgressRepository.findEagerByLearningGoalIdAndUserId(learningGoalId, user.getId());
+
+        if (learningGoalProgress.isPresent()) {
+            var lastModified = learningGoalProgress.get().getLastModifiedDate();
+            if (lastModified != null && lastModified.isAfter(Instant.now().minusSeconds(5))) {
+                // If we have updated the progress within the last five seconds, skip it
+                return;
```

Would be interesting for debugging purposes to log a message here instead of completely silently returning?